allennlp/modules/stacked_elmo_lstm.py

```diff
@@ -0,0 +1,168 @@
+            forward_layer = LSTMCellWithProjection(lstm_input_size,
+                                                   hidden_size,
+                                                   cell_size,
+                                                   go_forward,
+                                                   recurrent_dropout_probability,
+                                                   memory_cell_clip_value,
+                                                   state_projection_clip_value)
+            backward_layer = LSTMCellWithProjection(lstm_input_size,
+                                                    hidden_size,
+                                                    cell_size,
+                                                    not go_forward,
+                                                    recurrent_dropout_probability,
+                                                    memory_cell_clip_value,
+                                                    state_projection_clip_value)
+            lstm_input_size = hidden_size
+
+            self.add_module('forward_layer_{}'.format(layer_index), forward_layer)
+            self.add_module('backward_layer_{}'.format(layer_index), backward_layer)
+            forward_layers.append(forward_layer)
+            backward_layers.append(backward_layer)
+        self.forward_layers = forward_layers
+        self.backward_layers = backward_layers
+
+    def forward(self,  # pylint: disable=arguments-differ
+                inputs: PackedSequence,
+                initial_state: Optional[Tuple[torch.Tensor, torch.Tensor]] = None) -> Tuple[torch.Tensor,
+                                                                                            Tuple[torch.Tensor,
+                                                                                                  torch.Tensor]]:
+        """
+        Parameters
+        ----------
+        inputs : ``PackedSequence``, required.
+            A batch first ``PackedSequence`` to run the stacked LSTM over.
+        initial_state : Tuple[torch.Tensor, torch.Tensor], optional, (default = None)
+            A tuple (state, memory) representing the initial hidden state and memory
+            of the LSTM. Each tensor has shape (1, batch_size, output_dimension).
+
+        Returns
+        -------
+        output_sequence : ``torch.FloatTensor``
+            The encoded sequence of shape (num_layers, batch_size, sequence_length, hidden_size)
+        final_states: ``Tuple[torch.FloatTensor, torch.FloatTensor]``
+            The per-layer final (state, memory) states of the LSTM, with shape
+            (num_layers, batch_size, 2 * hidden_size) and  (num_layers, batch_size, 2 * cell_size)
+            respectively. The last dimension is duplicated because it contains the state/memory
+            for both the forward and backward layers.
+        """
+        if initial_state is None:
+            hidden_states: List[Optional[Tuple[torch.Tensor,
```

Comment:
Why is the type necessary here?  This doesn't actually do anything when running the code, right?

Reply:
No, it doesn't - however, it allows MyPy to run static type checking. I agree it's a bit annoying when the types are long like this, but the benefit is worth it for a large python codebase.

Reply:
Got it.